File: backend/src/main/java/at/ac/tuwien/sepr/groupphase/backend/service/ShoppingListService.java
package at.ac.tuwien.sepr.groupphase.backend.service;

import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.ShoppingItemDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.ShoppingItemSearchDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.ShoppingListDto;
import at.ac.tuwien.sepr.groupphase.backend.entity.DigitalStorageItem;
import at.ac.tuwien.sepr.groupphase.backend.entity.ShoppingItem;
import at.ac.tuwien.sepr.groupphase.backend.entity.ShoppingList;
import at.ac.tuwien.sepr.groupphase.backend.exception.AuthorizationException;
import at.ac.tuwien.sepr.groupphase.backend.exception.ConflictException;
import at.ac.tuwien.sepr.groupphase.backend.exception.ValidationException;

import java.util.List;
import java.util.Optional;

public interface ShoppingListService {

    /**
     * Validates and Creates a new {@link ShoppingItem} in the db.
     *
     * @param itemDto a shopping item without ID
     * @return an object of type {@link ShoppingItem} which is persisted and has an ID
     * @throws AuthorizationException if the user is not authorized to access a resource
     */
    ShoppingItem createShoppingItem(ShoppingItemDto itemDto) throws ValidationException, ConflictException, AuthorizationException;

    /**
     * Search for a shopping item in the database with given ID.
     *
     * @param itemId a valid ID
     * @return if the id exists in the DB, an Optional of a persisted ShoppingItem with given ID, an empty Optional otherwise
     * @throws AuthorizationException if the user is not authorized to access a resource
     */
    Optional<ShoppingItem> getShoppingItemById(Long itemId) throws AuthorizationException;

    /**
     * Search for a shopping list in the database with given ID.
     *
     * @param name a valid listName
     * @return if the id exists in the DB, an Optional of a persisted ShoppingList with given ID, an empty Optional otherwise
     */
    Optional<ShoppingList> getShoppingListByName(String name);

    /**
     * Search for a shopping list in the database with given ID.
     *
     * @param id a valid ID of a ShoppingList
     * @return if the id exists in the DB, an Optional of a persisted ShoppingList with given ID, an empty Optional otherwise
     */
    Optional<ShoppingList> getShoppingListById(Long id);


    /**
     * Search for a shopping list in the database with given ID.
     *
     * @param id a valid ID of a ShoppingList
     * @param itemSearchDto search parameters consisting of the product's name and its label's value
     * @return if the id exists in the DB, a List of a persisted ShoppingItems with the given ID, an empty Optional otherwise
     * @throws AuthorizationException if the user is not authorized to access a resource
     */
    List<ShoppingItem> getItemsByShoppingListId(Long id, ShoppingItemSearchDto itemSearchDto) throws AuthorizationException;

    /**
     * Create a new ShoppingList in the db.
     *
     * @param shoppingListDto a DTO of type shopping list ID null
     * @return an object of type {@link ShoppingList} which is persisted and has an ID
     */
    ShoppingList createList(ShoppingListDto shoppingListDto) throws ValidationException, ConflictException;

    /**
     * Delete a ShoppingItem from the db based on its ID.
     *
     * @param itemId a valid ID of a ShoppingItem
     * @return the deleted ShoppingItem
     * @throws AuthorizationException if the user is not authorized to access a resource
     */
    ShoppingItem deleteItem(Long itemId) throws AuthorizationException, ConflictException;

    /**
     * Delete a ShoppingList from the db based on its ID.
     *
     * @param shopId a valid ID of a ShoppingList
     * @return the deleted ShoppingList
     */
    ShoppingList deleteList(Long shopId) throws ValidationException, AuthorizationException, ConflictException;

    /**
     * Get all ShoppingLists from the db filtered by search parameters.
     *
     * @param searchParams name of the list, through which we search for it. Can also be null
     * @return a List of all persisted ShoppingLists
     * @throws AuthorizationException if the user is not authorized to access a resource
     */
    List<ShoppingList> getShoppingLists(String searchParams) throws AuthorizationException;

    /**
     * Transfer ShoppingItems to the server.
     *
     * @param items a List of ShoppingItemDto to be transferred
     * @return a List of DigitalStorageItem objects
     */
    List<DigitalStorageItem> transferToServer(List<ShoppingItemDto> items) throws AuthorizationException, ValidationException, ConflictException;

    /**
     * Validates and Updates a new {@link ShoppingItem} in the db.
     *
     * @param shoppingItemDto a DTO of type shopping item with existing ID
     * @return an object of type {@link ShoppingItem} which is updated
     * @throws AuthorizationException if the user is not authorized to access a resource
     * @throws ConflictException if there is a conflict with the persisted data
     * @throws ValidationException if the data in shoppingItemDto is not valid
     */
    ShoppingItem updateShoppingItem(ShoppingItemDto shoppingItemDto)
        throws ConflictException, ValidationException, AuthorizationException;

    /**
     * Deletes the specified chores.
     *
     * @param itemIds List of chore IDs to be deleted.
     * @return List of remaining chores after deletion.
     */
    List<ShoppingItem> deleteShoppingItems(List<Long> itemIds) throws AuthorizationException, ConflictException;
}

File: frontend/src/app/services/shopping-list.service.ts
import { Injectable } from '@angular/core';
import {HttpClient, HttpHeaders, HttpParams} from "@angular/common/http";
import {ItemDto, ShoppingItemDto, ShoppingItemSearchDto} from "../dtos/item";
import {Observable} from "rxjs";
import {ShoppingListDto} from "../dtos/shoppingList";
import {StorageItem} from "../dtos/storageItem";
import {AuthService} from "./auth.service";

@Injectable({
  providedIn: 'root'
})
export class ShoppingListService {

  private baseUri: string = 'http://localhost:8080/api/v1/shopping';

  constructor(private http: HttpClient,
              private authService: AuthService) {
  }

  /**
   * Creates a shopping item in the system
   *
   * @param item the data without ID for the shopping item that should be stored in the system
   * @return an Observable for the stored shopping list in the system
   */
  createItem(item: ShoppingItemDto): Observable<ShoppingItemDto> {
    return this.http.post<ItemDto>(this.baseUri, item);
  }

  /**
   * Find an existing shopping item in the system
   *
   * @param id the id of the shopping item that should already be stored in the system
   * @return an Observable for the existing shopping list in the system
   */
  getById(id: string): Observable<ShoppingItemDto> {
    return this.http.get<ShoppingItemDto>(this.baseUri + '/' + id);
  }

  /**
   * Find existing shopping items in the system
   *
   * @param shopId the id of the shopping list to which the shopping items are connected in the system
   * @param searchParams search parameters consisting of the products' name and their labels' value
   * @return an Observable for the existing shopping items in the system
   */
  getItemsWithShopId(shopId: string, searchParams: ShoppingItemSearchDto):Observable<ShoppingItemDto[]> {
    let params = new HttpParams();
    if (searchParams.productName) {
      params = params.append('productName', searchParams.productName);
    }
    if (searchParams.label) {
      params = params.append('label', searchParams.label);
    }
    return this.http.get<ShoppingItemDto[]>(this.baseUri + "/list-items/" + shopId, {params});
  }

  /**
   * Find an existing shopping list in the system
   *
   * @param shoppingListName the name of the list that should already be stored in the system
   * @return an Observable for the existing shopping list in the system
   */
  getShoppingListByName(shoppingListName: string): Observable<ShoppingListDto> {
    return this.http.get<ShoppingListDto>(this.baseUri + '/list/' + shoppingListName);
  }

  /**
   * Find an existing shopping list in the system
   *
   * @param shoppingListId the id of the list that should already be stored in the system
   * @return an Observable for the existing shopping list in the system
   */
  getShoppingListById(shoppingListId: string): Observable<ShoppingListDto> {
    return this.http.get<ShoppingListDto>(this.baseUri + '/listId/' + shoppingListId);
  }

  createList(list: ShoppingListDto): Observable<ShoppingListDto> {
    return this.http.post<ShoppingListDto>(this.baseUri + "/list-create", list);
  }

  deleteItem(itemId: number): Observable<ShoppingItemDto> {
    return this.http.delete<ShoppingItemDto>(this.baseUri + '/' + itemId);
  }

  deleteList(shopId: string): Observable<ShoppingListDto> {
    return this.http.delete<ShoppingListDto>(this.baseUri + '/delete/' + shopId);
  }

  /**
   * Get all shopping lists filtered by search parameters
   *
   * @param searchParams the search parameters consisting only of one string representing the lists' name
   * @return on Observable for the array of lists, which fulfil the search criteria
   */
  getShoppingLists(searchParams: string): Observable<ShoppingListDto[]> {
    let params = new HttpParams();
    if (searchParams) {
      params = params.append('searchParams', searchParams);
    }
    return this.http.get<ShoppingListDto[]>(this.baseUri + '/lists', {params});
  }

  transferToStorage(shoppingItems: ShoppingItemDto[]): Observable<StorageItem[]> {
    return this.http.post<StorageItem[]>(this.baseUri + '/storage', shoppingItems);
  }

  /**
   * Update an item in the system.
   *
   * @param item the data for the item that should be updated
   * @return an Observable for the updated item
   */
  updateItem(item: ShoppingItemDto): Observable<ShoppingItemDto> {
    return this.http.put<ShoppingItemDto>(`${this.baseUri}/${item.itemId}`, item);
  }

  deleteItems(items: ShoppingItemDto[]) {
    console.log(items)
    const itemIds = items.map(item => item.itemId);
    return this.http.delete<ShoppingItemDto[]>(this.baseUri + '/delete', {params: {itemIds: itemIds.join(',')}});
  }

}

